source/blog/2015-05-08-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,216 @@
+author: Sam Phippen
+---
+
+##Warmup: the `===` protocol
+
+In Ruby, a number of objects implement the `===` protocol. Some of the objects
+that implement this protocol include:
+
+* all class objects
+* lambdas
+* regexes
+* ranges
+
+Conceptually, this protocol represents a "match". That is, the object that
+implements the `===` declares whether or not it "matches" the other object.
+For example, `Array === []` returns `true`. This is because an empty array is a
+subclass of the array class. `Array === {}` returns false because a hash is not
+a subclass of array. For regexes, `===` returns true if the passed string
+matches the regex, and false if it does not.
+
+The `===` method is used in Ruby case statements to decide if a branch should
+be taken. For example
+
+```
+case 3
+when (1...18)
+    puts "hello"
+end
+```
+
+will print hello, because `(1...18) === 3` returns `true`.
+
+The new RSpec matchers make heavy use of `===` internally, if you'd like to get
+more of an understanding of `===`, I'd heavily suggest you pop open IRB and
+play around with it a little before moving on.
+
+## RSpec matchers
+
+In an RSpec expectation expression like `expect(object_1).to eq(object_2)`
+there are two primary objects at play, the **target** and the **matcher**. The
+**target** object is created by "expect" and basically wraps the passed object
+to it in a class called `ExpectationTarget`. The primary methods on the
+`ExpectationTarget` that users see are `to` and `not_to`/`to_not`. Let's focus
+on `to`.
+
+The `to` method is basically responsible for invoking the `matches?` method on the
+passed matcher. It's possible to see this happening quite easily in an IRB session.
+
+```
+>> require "rspec/expectations"
+=> true
+>> o = Object.new
+=> #<Object:0x007fa20d221490>
+>> o.extend(RSpec::Matchers)
+=> #<Object:0x007fa20d221490>
+>> o.expect(3).to Object.new
+NoMethodError: undefined method `matches?' for #<Object:0x007fa20d1f4e68>
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:27:in `with_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:54:in `to'
+    from (irb):4
+    from /Users/sam/.rubies/ruby-2.1.5/bin/irb:11:in `<main>'
+>>
+```
+
+In this session, we extend the RSpec DSL on to a blank object and then call
+`o.expect(3).to Object.new`. This blows up because our new object does not implement
+the `matches?` method.
+
+In RSpec expectations, most matchers inherit from a class called `BaseMatcher`
+which causes `matches?` to delegate to a method called `match`. Some matchers have
+a very simple definition of `match`. For example, here's the definition of `match`
```

Comment:
The fact that RSpec's built-in matchers subclass a base class which defines `matches?` to delegate to `match` doesn't seem important here.  `BaseMatcher` is definitely an implementation detail and is not part of RSpec's public API or semantics.  You also risk confusing the reader with the mention of `matches?` and `match` -- the fact that we've chosen to reduce duplication using that implementation mechanism has nothing to do with how composable matchers works but its presence here in the blog post might suggest that it does.  It's also confusing that the `matches?` protocol takes one arg (`actual`) but `match` takes two.

Instead, I'd recommend just talking about the matcher protocol (`matches?`) without reference to any extra layers in the built-in matchers such as `BaseMatcher` or `match`.  If you want to show what the definition is for `eq`, you could say something like;

> For example, here's the `matches?` logic for the `eq` matcher:
>
> ``` ruby
> def matches?(actual)
>   actual == expected
> end
> ```